### ISSUE TITLE
support `import.meta.dirname` and `import.meta.filename`

### ISSUE THREAD

**Issue by EricMCornelius** [webpack/webpack#18320]
# Bug report
#17931 introduced functionality to support __dirname and __filename shims for module output.

Unfortunately, there are false assumptions in the behavior - namely any import.meta.url that has a query param will break.

https://github.com/webpack/webpack/pull/17931/files#diff-ec15f6ea9baa75e9b3ca10449f3b9b7e04e56828330aa5a31d8a83aadd733ed0R225

https://github.com/webpack/webpack/blame/e97af9b5317bc0e7fdbc035b98a577edfe258b83/lib/NodeStuffPlugin.js#L230

**What is the current behavior?**

Current behavior just string appends `'/..'` to the `import.meta.url` value to try to get a directory.  This is inaccurate behavior for any import urls which contain query params.

**If the current behavior is a bug, please provide the steps to reproduce.**

Don't have an easy example I can generate offhand here, but any webpack node build with `__dirname: 'node-module'` is prone to this issue when running, e.g., `const x = import('./relative_path.mjs?reload=123')`;

**What is the expected behavior?**

**Other relevant information:**
webpack version:  6.91.0
Node.js version: 20.11.1
Operating System: linux/windows x64
Additional tools:

**Comment by alexander-akait** [comment]
I want to say it is expected, `__dirname` !== "import.meta.url" (with `/..`), so that is why you don't query, `__dirname` === `import.meta.dirname` (node.js 21 and we will improve it soon, if you browserslist contains node 21)

Try this:

**index.mjs**

```js
const x = import('./test.mjs?reload=123');
```

**test.mjs**

```js
console.log(import.meta.url);
console.log(import.meta.dirname);
```

**Comment by alexander-akait** [comment]
/cc @vankop I think we need to start support `import.meta.dirname` and `import.meta.filename`, there are two improvements:
1. Use `import.meta.dirname` and `import.meta.filename` instead `__webpack_fileURLToPath__(import.meta.url + "/..")` if `target`/`browserslist` support it (if it is node target or node like)
2. Don't output `undefined` for `import.meta.dirname` and `import.meta.filename`, i.e. output them as is if they supported or replace on our shim with `__webpack_fileURLToPath__(import.meta.url + "/..")` (I think we need options here i.e. `node.importMetaDirname` and `node.importMetaFilename`)

**Comment by inoyakaigor** [comment]
For anyone who need a temporarily solution.

Change 
```js
import.meta.dirname
```
to
```js
path.dirname(process.argv[1]) //*
```
`*` In my case there is a script path of my nodejs app. In your case there can be another file path

**Comment by alexander-akait** [comment]
bump

**Comment by Clashsoft** [comment]
bump

**Comment by AntonioGargaro** [comment]
bump

**Comment by kane50613** [comment]
bump

**Comment by Schweinepriester** [comment]
inactivity != not relevant anymore, please reopen

**Comment by inoyakaigor** [comment]
@alexander-akait please reopen

**Comment by robahtou** [comment]
bump 🤜🏽

**Comment by fpapado** [comment]
Hi @alexander-akait, we recently ran into an issue with `import.meta.dirname` being undefined, and I'm interested in contributing some of the changes proposed in this thread 😌 

I glanced through the relevant codepaths (`lib/config/defaults.js`, `lib/dependencies/ImportMetaPlugin.js`, `lib/NodeStuffPlugin.js` and the test setup), and I have a sketch for how to implement this.

However, before I go down that path, I want to recap my understanding of the specification 📝  As you mentioned earlier, we will need options for `node.importMetaDirname` and `node.importMetaFilename`.

**These options seem applicable for the first phase:**
  - `'none'` or `'unsupported'`: the current default in v5. Emits these properties as undefined, with a webpack comment.
  - `false` or `'node-module'`: Webpack won't touch the `import.meta.filename`/`import.meta.dirname` code, leaving the default Node.js behavior. The dirname of the **output** file when run in a Node.js environment.
    - If the browserslist targets do not support these properties, we use an `import.meta.url` shim.
    - **Question**: Should this be the future default, when `output.module` is enabled? Can be opted-in manually, or with `futureDefaults`.

**This could be in a second phase:**
  - `true` or `'resolved-in-context'`: The dirname/filename of the input file relative to the [context option](https://webpack.js.org/configuration/entry-context/#context).
    - This is symmetric with the `__filename` and `__dirname` `true` option, because I can imagine people wanting to resolve based on the input files.
    - **Question**: Should this support both CJS and ESM output?

**And a third phase, if there is a need:**
- `node-cjs`: Replace `import.meta.filename` / `import.meta.dirname` in ESM modules with `__filename` / `__dirname` equivalents when emitting CJS.
  - This would be the inverse of the `node.__filename` / `node.__dirname` `node-module` option. I have not seen much need for this at the moment, but let me know if it makes sense.

I wrote the names to be analogous to [node.__dirname](https://webpack.js.org/configuration/node/#node__filename) and [node.__filename](https://webpack.js.org/configuration/node/#node__dirname), but I personally find the `true` and `false` values confusing. So I wrote some suggestions about stringy values, but we can bikeshed that if you prefer 😌 

--- 

Aside: While implementing `import.meta.filename` / `import.meta.dirname` for ESM would help remove the shim and thus fix the original issue in this thread, I think the fix itself could be done in a follow-up.

**Comment by fpapado** [comment]
I also wonder whether the current shim could be fixed by going via `new URL()` instead of appending to the path:

**index.mjs**

```js
const x = import('./test.mjs?reload=123');
```

**test.mjs**

```js
import {fileURLToPath} from 'node:url';

/* file:///Users/fotis/oss/dirname-test/test.mjs?reload=123 */
console.log('native import.meta.url', import.meta.url);

/* /Users/fotis/oss/dirname-test */
console.log('native import.meta.dirname', import.meta.dirname);

/* /Users/fotis/oss/dirname-test/test.mjs */
console.log('shim dirname', fileURLToPath(import.meta.url + "/.."))

/* /Users/fotis/oss/dirname-test/ */
console.log('shim dirname alternative', fileURLToPath(new URL("./", import.meta.url)))
```

In any case, I'm more interested in native `import.meta.dirname` / `import.meta.filename` support personally, I just mention this for posterity and my own understanding of webpack's problem space 💭

**Comment by alexander-akait** [comment]
Yeah, it can be replaced by `fileURLToPath`, to be honestly webpack will generate the same code for back compatibility for Node.js target where `import.meta.dirname` and `import.meta.url` are not supported, for browsers we should generate a warning by default but allows to set an option to generate `import.meta.url`, i.e. logic will be same as for https://webpack.js.org/configuration/node/#node__filename and https://webpack.js.org/configuration/node/#node__dirname, this task is not hard, so if someone want to finish it as soon as possible - PR welcome
